crates/cli-support/src/js/mod.rs

```diff
@@ -842,7 +847,43 @@ impl<'a> Context<'a> {
         Ok(())
     }
 
-    fn gen_init(&mut self, module_name: &str, needs_manual_start: bool) -> String {
+    fn ts_for_init_fn(has_memory: bool) -> &'static str {
+        macro_rules! initial_ts {
+            () => ("
+                \n\
+                /**\n\
+                * If `module_or_path` is {RequestInfo}, makes a request and\n\
+                * for everything else, calls `WebAssembly.instantiate` directly.\n\
+                *\n\
+                * @param {RequestInfo | BufferSource | WebAssembly.Module} module_or_path\n\
+            ");
+        }
+        if has_memory {
+            concat!(initial_ts!(), "\
+            * @param {WebAssembly.Memory} maybe_memory\n\
+            *\n\
+            * @returns {Promise<WebAssembly.ResultObject | WebAssembly.Instance>}\n\
+            */\n\
+            export function init (\
+                module_or_path: RequestInfo | BufferSource | WebAssembly.Module,\
+                maybe_memory: WebAssembly.Memory\
+            ): Promise<WebAssembly.ResultObject | WebAssembly.Instance>;
```

Comment:
On which conditions does it return `WebAssembly.Instance` as opposed to `ResultObject`?

Reply:
`WebAssembly.instantiate` has two overloading methods ([1](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/instantiate), [2](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/8c9b0319e58804090687ba85e15e6ac909c8452e/types/webassembly-js-api/index.d.ts#L113-L114)). Maybe define those overloads in the definition itself?

Reply:
Oh this doesn't actually resolve to either, but it sort of resolves to this module itself. It's probably fine to just say it resolves to `any` for now and we can refine it necessary if need be.

(it definitely doesn't result in `Instance` or `ResultObject` though)